### DISCUSSION TITLE
How to work Glyph-cache in proxy

### DISCUSSION THREAD

**Discussion by tao363** [FreeRDP/FreeRDP#10397] [Q&A] [answered]
Hello, I would like to ask a question, how to make glyph-cache work in proxy mode? What do I need to do about it?

**Comment by akallabeth** [comment, marked as answer]
* `pf_client_pre_connect` is the place we deactivate (most) of the legacy `GDI` stuff due to compatibility issues.
* https://github.com/FreeRDP/FreeRDP/discussions/10347
* why do you want it in the first place, this sounds like a configuration issue

**Reply by tao363** [reply]
Thank you very much for your quick answer！！

＞ Do I need to use freerdp3 to use glyph cache in proxy mode?
＞ I want to record the text content displayed by the user interface in proxy mode instead of the entire frame

**Reply by akallabeth** [reply]
> Do I need to use freerdp3 to use glyph cache in proxy mode?

yes, it was a complete rewrite from 2.0 and an update is highly recommended.

> I want to record the text content displayed by the user interface in proxy mode instead of the entire frame

will not work. There is no decoding (text/frame/...) on the proxy, you need to dump/record the `RDP` data stream and do the decoding on a separate client.

Even if you try to continue on this road, the whole thing depends entirely on the client settings and various modes of graphics representation might be used:
* legacy GDI, the thing you are most likely looking at
  * bitmap data (the thing in legacy GDI that is used by modern operating systems as most data can no longer be redirected with GDI calls
* [MS-RDPRFX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdprfx/62495a4a-a495-46ea-b459-5cde04c44549) the windows 7 style graphics redirection
* [MS-RDPEGFX](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegfx/da5c75f9-cd99-450c-98c4-014a496942b0) the windows 8 and newer style graphics redirection

**Reply by tao363** [reply]
Ok, thank you for your reply. Session playback requires a certificate that is installed on the particular computer, right?

**Reply by akallabeth** [reply]
no, the proxy has access to the data inside the `TLS` tunnels.
The `RDP` connection terminates there, so the proxy is more or less man in the middle.
your connection looks like this: ` RDP-client <-- [TLS] --> proxy <-- [TLS] --> target RDP server`

**Reply by tao363** [reply]
Well. I think I know what you mean. You can dump the package from the server to the client on the proxy, save it, and play it back on another client. Is that what it means?

**Reply by akallabeth** [reply]
yes.
you need to extract the client settings though (or you know the connection settings) as the initially sent configuration is only read by the server component and not in the data sent to the client.

**Reply by tao363** [reply]
OK, thank you very much for your prompt answer!
